### PR TITLE
DAOS-10148 tse: fix TSE task buffer to account for aarch64 architecture

### PR DESCRIPTION
**Pull request by mchaarawi**
pthread_mutex_t on arm is 8 bytes longer than on x86, so adjust the TSE
preallocated buffer to fix the static compile time assertion.

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]
This is probably fine but would it make sense to do something like

TSE_TASK_ARG_LEN (872 + sizeof(pthread_mutex_t)) ?

**Review by wiliamhuang — APPROVED** [review]

File: src/include/daos/tse.h
Comment (liuxuezhao) [inline src/include/daos/tse.h:31]: I'm afraid only changing is not enough, we may ask the person who reported the bug to test it on arm node.
The reason is TSE_TASK_SIZE, TSE_PRIV_SIZE, TSE_TASK_ARG_LEN are calculated, and with such compiling assertion in tse.c -
D_CASSERT(sizeof(struct tse_task) == TSE_TASK_SIZE);                                                    
D_CASSERT(sizeof(struct tse_task_private) <= TSE_PRIV_SIZE);

If only enlarge TSE_TASK_ARG_LEN for arm, looks possibly break the original assumption.

**Review by liuxuezhao — CHANGES_REQUESTED** [review]
I thin need test on arm before land, please see the comment.

File: src/include/daos/tse.h
Comment (mchaarawi) [reply src/include/daos/tse.h:31]: yea you're right we need to adjust those. let me fix that.

Comment (mchaarawi) [reply src/include/daos/tse.h]: fixed. also took Jeff's suggestion and extended it to the other variables.
i agree we need him to test this to verify as i mentioned in the ticket, but should get PR approval first.

File: src/include/daos/tse.h
Comment (liuxuezhao) [inline src/include/daos/tse.h:19]: I just feel this looks wired, why need to tight task arg len with size of mutex, at lest need details comments, and looks not very good to add such thing here for a very common used task length definition. (the size problem on arm is hit on at cli_obj.c)

Another option is at cli_obj.c reduce "OBJ_TGT_INLINE_NR" for ARM platform with comment explain the size difference of mutex used in "struct obj_reasb_req" on ARM platform. 
just FYI.

**Review by liuxuezhao — APPROVED** [review]
just want to remove the request change, although the current change of "TSE_TASK_SIZE" looks a kind of not very good to me.

File: src/include/daos/tse.h
Comment (mchaarawi) [reply src/include/daos/tse.h:19]: Well the issue is that originally this didn’t look good as it was hardcoded. So this is just making it more dynamic as the pthread mutex is of different size. I can add a comment that pthread mutex is of different size on different platforms to make this clearer. I don’t have other suggestions. Changing OBJ_TGT_INLINE_NR is just doing the same thing really in a different place.

Comment (mchaarawi) [reply src/include/daos/tse.h:19]: if you prefer i can revert to the original change adding the:
#ifdef __aarch64__
and just hardcode things.
but i was just addressing the suggestion from @jolivier23 to just use sizeof()

Comment (liuxuezhao) [reply src/include/daos/tse.h:19]: hi, i just feel looks not very easy to understand what's the reason of "#define TSE_TASK_ARG_LEN	(840 + sizeof(pthread_mutex_t))" in tse.h, as some tasks even don't use any mutex.
As the real problem is introduced by cli_obj.c's compiling, probably fine to fix it there.
But not a big deal, can leave as it and later after we cleanup things in cli obj we may change it back. thanks.

**Review by liuxuezhao — DISMISSED** [review]

**Review by jolivier23 — DISMISSED** [review]

File: src/include/daos/tse.h
Comment (liuxuezhao) [reply src/include/daos/tse.h:19]: the suggestion is to only change the task in cli_obj.c which actually caused the compiling failure, to avoid confuse to change a very commonly used task struct's length.

Comment (mchaarawi) [reply src/include/daos/tse.h:19]: im a little against the change in cli_obj.c because it would change the behavior on the client between 2 different architectures. we can change it for both, but in this case for x86, then the task buf size is not in excess. and if someone adds more fields later and forgets to bump the task buf size, it would still work on x86 but would break aarch64. so i think this needs to be account for in the task buf size.

**Review by wiliamhuang — APPROVED** [review]

**Comment by mchaarawi** [comment]
Ticket reporter mentioned the PR fixes the build issues.

**Pull request merged**
